flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/utils/SqlValidatorUtils.java

```diff
@@ -76,9 +80,25 @@ private static void adjustTypeForMultisetConstructor(
     }
 
     private static SqlNode castTo(SqlNode node, RelDataType type) {
-        return SqlStdOperatorTable.CAST.createCall(
-                SqlParserPos.ZERO,
-                node,
-                SqlTypeUtil.convertTypeToSpec(type).withNullable(type.isNullable()));
+        final SqlDataTypeSpec dataTypeSpec;
+        if (SqlTypeUtil.isMap(type)) {
```

Comment:
Isn't this only fixing the top-level problem but not nested problems such as `ARRAY[ARRAY[MAP]]` or even more complex `ARRAY[MAP[ARRAY[MAP]]]`?

Reply:
Btw in `org.apache.flink.table.planner.functions.inference.TypeInferenceOperandChecker#castTo` we would have a similar issue, correct?